extended-utxo-spec/extended-utxo-specification.tex

```diff
@@ -249,12 +240,94 @@ \subsection{Structure of the document}
+
+\begin{ruledfigure}{H}
+  \center{
+    \begin{tabular}{rl}
+      \B&: the type of booleans, $\{\false, \true\}$\\
+      \N&: the type of natural numbers, $\{0, 1, 2, \ldots\}$\\
+      \Z&: the type of integers, $\{,\ldots, -2, -1, 0, 1, 2, \ldots\}$\\
+      \H&: the type of \textit{bytestrings}, $\bigcup_{n=0}^{\infty}\{0,1\}^{8n}$\\
+      $\qty = \N$&: an amount of currency, always non-negative\\
+      $\qtypm = \Z$&: a possibly negative amount of currency\\
+      $\s{SlotNumber} = \N$&: a slot number\\
+      $\s{Address} = \N$&: the address of an object in the blockchain\\
+      $\s{TxId} = \N$&: the identifier of a previous transaction on the chain\\
+    \end{tabular}
+  }
+  \caption{Basic types}
+  \label{fig:basic-types}
+\end{ruledfigure}
+
+\noindent We regard $\N$ as a subtype of $\Z$ and convert freely between
+natural numbers and non-negative integers.  A bytestring is a sequence
+of 8-bit bytes.
+
+\medskip
+\noindent In practice an \textsf{Address} will be a
+collision-resistant hash of some object (so the address of an object
+$X$ is $X^{\#}$ in the notation of Figure~\ref{fig:basic-notation}),
+and the blockchain will provide an efficient way to retrieve the
+original object given its hash.  Compiled Plutus Core scripts are
+stored and transmitted in a serialised form, and when we talk about
+the hash of a Plutus Core script we mean the hash of the serialised
+version.
+
+% Accessing objects indirectly via addresses is helpful because it can
+% help to reduce memory and disk usage: for example, there may be
+% scripts for common validation scenarios used in many transactions, and
+% it is more efficient to store single copies of such scripts rather
+% than having hundreds of transactions each with their own copy.
+
+\medskip
+\noindent We assume that every transaction in a ledger has a unique identifier
+of type \s{TxId}: in an implementation this would probably be the
+hash of the transaction.
+
+\subsection{The \Data{} type}
+We also define a type \Data{} which can be used to pass information
+into scripts in a type-safe manner: see Figure~\ref{fig:data-defn}. The
+definition is given here in EBNF form, but can easily be translated to
+a Haskell type, for instance.
+
+\begin{ruledfigure}{H}
+\begin{alltt}
+  \Data =
+      "I" \(\N\)
+    | "B" \(\H\)
+    | "Constr" \(\N (\List{\Data})\)
+    | "Map" \(\List{\Data\times\Data}\)
+\end{alltt}
+\caption{The \Data{} type}
+\label{fig:data-defn}
+\end{ruledfigure}
+
+\noindent Thus values of type \Data{} are nested sums and products built up
+recursively from the base types of integers and bytestrings. This
+allows one to encode a large variety of first-order data structures:
+for example, we could encode the list [11,22,33] as \texttt{Constr 1
+  (I 11, Constr 1 (I 22, Constr 1 (I 33, Constr 0)))}, using
+\texttt{Constr 0} to represent the \texttt{Nil} constructor of a list
+type, and \texttt{Constr 1} to represent the \texttt{Cons}
+constructor.
```

Comment:
FYI I'm planning to add lists as a primitive constructor since otherwise they're pretty inefficient. And CBOR supports them.

Reply:
OK. I've added that and changed the example to use Maybe instead of List.